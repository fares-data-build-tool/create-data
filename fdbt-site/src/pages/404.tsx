import React, { ReactElement } from 'react';
import Layout from '../layout/Layout';
import { FEEDBACK_LINK } from '../constants';

const title = 'Page not found - Fares data build tool';
const description = 'Page not found page of the Fares data build tool';

const PageNotFound = (): ReactElement => (
    <Layout title={title} description={description}>
        <main className="govuk-main-wrapper app-main-class" id="main-content" role="main">
            <div className="govuk-grid-row">
                <div className="govuk-grid-column-two-thirds">
                    <div>
                        <h1 className="govuk-heading-xl">Page not found</h1>
                        <p className="govuk-body">If you typed the web address, check it is correct.</p>
                        <p className="govuk-body">
                            If you pasted the web address, check you copied the entire address.
                        </p>
                        <p className="govuk-body">
                            If the web address is correct or you selected a link or button,{' '}
                            <a className="govuk-link" id="feedback_link" href={FEEDBACK_LINK}>
                                contact
                            </a>{' '}
                            us about your fares data.
                        </p>
                    </div>
                    <br />
                    <a
                        href="/operator"
                        role="button"
                        draggable="false"
                        className="govuk-button"
                        data-module="govuk-button"
                    >
                        Start again
                    </a>
                </div>
            </div>
        </main>
    </Layout>
);

export default PageNotFound;
