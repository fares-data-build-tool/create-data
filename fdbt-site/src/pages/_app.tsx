import '../design/Pages.scss';
import '../design/Layout.scss';
import '../design/Components.scss';
import { AppProps } from 'next/app';
import React, { ReactElement, useEffect } from 'react';

const MyApp = ({ Component, pageProps }: AppProps): ReactElement => {
    useEffect(() => {
        document.getElementsByTagName('body')[0].classList.add('js-enabled');
    });

    // eslint-disable-next-line react/jsx-props-no-spreading
    return <Component {...pageProps} />;
};

export default MyApp;
