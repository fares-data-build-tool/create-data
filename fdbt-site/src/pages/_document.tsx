import Document, { Html, Head, Main, NextScript, DocumentInitialProps } from 'next/document';
import React, { ReactElement } from 'react';
import { parseCookies } from 'nookies';
import Header from '../layout/Header';
import { COOKIES_POLICY_COOKIE, COOKIE_PREFERENCES_COOKIE, ID_TOKEN_COOKIE } from '../constants';
import { DocumentContextWithSession } from '../interfaces';
import { getCsrfToken, ResponseWithLocals } from '../utils';
import { CookieBannerMessage } from '../layout/CookieBanner';

interface DocumentProps extends DocumentInitialProps {
    nonce: string;
    isAuthed: boolean;
    csrfToken: string;
    url: string;
    showCookieBanner: boolean;
    allowTracking: boolean;
}

class MyDocument extends Document<DocumentProps> {
    static async getInitialProps(ctx: DocumentContextWithSession): Promise<DocumentProps> {
        const initialProps = await Document.getInitialProps(ctx);
        const nonce = (ctx.res as ResponseWithLocals)?.locals?.nonce ?? null;
        const csrfToken = getCsrfToken(ctx);

        if (ctx.pathname !== '/') {
            ctx.res?.setHeader('X-Robots-Tag', 'none, noindex, nofollow, noimageindex, noarchive');
        }

        const url = ctx.req?.url ?? '';

        const cookies = parseCookies(ctx);
        const idTokenCookie = cookies[ID_TOKEN_COOKIE];

        const cookiePreferencesSet = cookies[COOKIE_PREFERENCES_COOKIE]
            ? JSON.parse(cookies[COOKIE_PREFERENCES_COOKIE])
            : false;
        const showCookieBanner = !cookiePreferencesSet && url !== '/cookies';
        const allowTracking = cookies[COOKIES_POLICY_COOKIE] ? JSON.parse(cookies[COOKIES_POLICY_COOKIE]).usage : false;

        return {
            ...initialProps,
            nonce,
            isAuthed: !!idTokenCookie,
            csrfToken,
            url,
            showCookieBanner,
            allowTracking,
        };
    }

    render(): ReactElement {
        return (
            <Html lang="en" className="govuk-template app-html-class flexbox no-flexboxtweener">
                <Head nonce={this.props.nonce}>
                    {process.env.STAGE === 'prod' && (
                        <>
                            <script
                                async
                                src="https://www.googletagmanager.com/gtag/js?id=UA-173062045-1"
                                nonce={this.props.nonce}
                            />
                            <script
                                nonce={this.props.nonce}
                                // eslint-disable-next-line react/no-danger
                                dangerouslySetInnerHTML={{
                                    __html: `window['ga-disable-UA-173062045-1'] = ${!this.props.allowTracking};
                                        window.dataLayer = window.dataLayer || [];
                                        function gtag(){dataLayer.push(arguments);}
                                        gtag('js', new Date());
                                        gtag('config', 'UA-173062045-1', {'anonymize_ip': true});`,
                                }}
                            />
                        </>
                    )}
                </Head>
                <body className="govuk-template__body app-body-class">
                    <a href="#main-content" className="govuk-skip-link">
                        Skip to main content
                    </a>
                    <div id="js-cookie-banner" />

                    {this.props.showCookieBanner ? (
                        <div className="no-js-cookie-banner">
                            <CookieBannerMessage />
                        </div>
                    ) : null}

                    <Header isAuthed={this.props.isAuthed} csrfToken={this.props.csrfToken} />
                    <Main />
                    <NextScript nonce={this.props.nonce} />
                    <script src="/scripts/all.js" nonce={this.props.nonce} />
                    <script nonce={this.props.nonce}>window.GOVUKFrontend.initAll()</script>
                </body>
            </Html>
        );
    }
}

export default MyDocument;
