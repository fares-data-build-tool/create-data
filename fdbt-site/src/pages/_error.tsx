import '../design/Pages.scss';
import React, { ReactElement } from 'react';
import { NextPageContext } from 'next';
import Layout from '../layout/Layout';

const title = 'Error - Fares data build tool';
const description = 'Error page of the Fares data build tool';

interface ErrorProps {
    statusCode: number;
}

const Error = ({ statusCode }: ErrorProps): ReactElement => (
    <Layout title={title} description={description}>
        <main className="govuk-main-wrapper app-main-class" id="main-content" role="main">
            <div className="govuk-grid-row">
                {!statusCode || statusCode !== 404 ? (
                    <div>
                        <h1 className="govuk-heading-xl">Sorry, there is a problem with the service.</h1>
                        <p className="govuk-body">Try again later.</p>
                        <p className="govuk-body">Your answers have not been saved, click continue to start again.</p>
                        <p className="govuk-body">
                            Contact us with the feedback button at the top of this page for assistance.
                        </p>
                    </div>
                ) : (
                    <div>
                        <h1 className="govuk-heading-xl">Page not found</h1>
                        <p className="govuk-body">If you entered a web address please check it was correct.</p>
                        <p className="govuk-body">Click continue to start again.</p>
                    </div>
                )}

                <br />
                <a
                    href="operator"
                    role="button"
                    draggable="false"
                    className="govuk-button govuk-button--start"
                    data-module="govuk-button"
                >
                    Continue
                </a>
            </div>
        </main>
    </Layout>
);

Error.getServerSideProps = (ctx: NextPageContext): {} => {
    return { statusCode: ctx?.res?.statusCode };
};

export default Error;
