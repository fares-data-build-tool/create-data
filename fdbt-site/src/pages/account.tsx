import React, { ReactElement } from 'react';
import { NextPageContext } from 'next';
import { parseCookies } from 'nookies';
import TwoThirdsLayout from '../layout/Layout';
import { ID_TOKEN_COOKIE } from '../constants';
import { getNocFromIdToken, getAttributeFromIdToken } from '../utils';

const title = 'Account Details - Create Fares Data Service';
const description = 'Account Details page of the Create Fares Data Service';

interface AccountDetailsProps {
    emailAddress: string;
    nocCode: string;
}

const AccountDetails = ({ emailAddress, nocCode }: AccountDetailsProps): ReactElement => {
    const passwordDots: ReactElement[] = [];
    for (let i = 0; i < 8; i += 1) {
        passwordDots.push(<span key={i} className="dot" />);
    }
    return (
        <TwoThirdsLayout title={title} description={description}>
            <div className="account-details-page">
                <h1 className="govuk-heading-xl">Account Details</h1>

                <div className="content-wrapper">
                    <p className="govuk-body govuk-!-font-weight-bold content-one-quarter">Email Address</p>
                    <p className="govuk-body email-content content-three-quarters">{emailAddress}</p>
                </div>
                <div className="content-wrapper">
                    <p className="govuk-body govuk-!-font-weight-bold content-one-quarter">Password</p>
                    <span className="password-dots">{passwordDots}</span>
                    <span className="change-password">
                        <a
                            href="/changePassword"
                            role="button"
                            draggable="false"
                            className="govuk-button"
                            data-module="govuk-button"
                            id="change-password-button"
                            aria-label="change password"
                        >
                            Change
                        </a>
                    </span>
                </div>
                <div className="content-wrapper">
                    <p className="govuk-body govuk-!-font-weight-bold content-one-quarter">Operator</p>
                    <p className="govuk-body content-three-quarters">{nocCode.replace('|', ', ')}</p>
                </div>
            </div>
            <a
                href="/home"
                role="button"
                draggable="false"
                className="govuk-button"
                data-module="govuk-button"
                id="home-button"
            >
                Home
            </a>
        </TwoThirdsLayout>
    );
};

export const getServerSideProps = (ctx: NextPageContext): { props: AccountDetailsProps } => {
    const cookies = parseCookies(ctx);
    if (!cookies[ID_TOKEN_COOKIE]) {
        throw new Error('Necessary cookies not found to show account details');
    }
    const noc = getNocFromIdToken(ctx);
    const email = getAttributeFromIdToken(ctx, 'email');

    if (!email || !noc) {
        throw new Error('Could not extract the user email address and/or noc code from their ID token');
    }
    return { props: { emailAddress: email, nocCode: noc } };
};

export default AccountDetails;
