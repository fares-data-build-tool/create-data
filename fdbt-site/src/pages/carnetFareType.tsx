import React, { ReactElement } from 'react';
import TwoThirdsLayout from '../layout/Layout';
import { ErrorInfo, NextPageContextWithSession, OperatorAttribute } from '../interfaces';
import CsrfForm from '../components/CsrfForm';
import ErrorSummary from '../components/ErrorSummary';
import FormElementWrapper from '../components/FormElementWrapper';
import FareTypeRadios from '../components/FareTypeRadios';
import { getSessionAttribute } from '../utils/sessions';
import { CARNET_FARE_TYPE_ATTRIBUTE, OPERATOR_ATTRIBUTE, FARE_TYPE_ATTRIBUTE } from '../constants/attributes';
import { redirectTo } from './api/apiUtils';
import { isSchemeOperator, getCsrfToken } from '../utils';
import { isFareTypeAttributeWithErrors } from '../interfaces/typeGuards';

const title = 'Carnet Fare Type - Create Fares Data Service ';
const description = 'Carnet Fare Type selection page of the Create Fares Data Service';

const errorId = 'fare-type-single';

interface CarnetFareTypeProps {
    operatorName: string;
    errors: ErrorInfo[];
    csrfToken: string;
}

const carnetRadioProps = [
    {
        fareType: 'single',
        label: 'Single, point to point carnet',
        hint: 'A ticket bundle of various point to point journeys',
    },
    {
        fareType: 'return',
        label: 'Return carnet',
        hint: 'A ticket bundle of inbound and outbound journeys',
    },
    {
        fareType: 'flatFare',
        label: 'Flat fare carnet',
        hint: 'A ticket bundle of flat fare journeys',
    },
    {
        fareType: 'period',
        label: 'Period carnet',
        hint: 'A ticket bundle where each ticket is valid for a number of days, weeks, months or years',
    },
    {
        fareType: 'multiOperator',
        label: 'Multi-operator carnet',
        hint: 'A ticket bundle that covers more than one operator',
    },
    {
        fareType: 'schoolService',
        label: 'School service carnet',
        hint: 'A ticket bundle available to pupils in full-time education',
    },
];

const CarnetFareType = ({ operatorName, errors = [], csrfToken }: CarnetFareTypeProps): ReactElement => {
    return (
        <TwoThirdsLayout title={title} description={description} errors={errors}>
            <CsrfForm action="/api/carnetFareType" method="post" csrfToken={csrfToken}>
                <>
                    <ErrorSummary errors={errors} />
                    <div className={`govuk-form-group ${errors.length > 0 ? 'govuk-form-group--error' : ''}`}>
                        <fieldset className="govuk-fieldset" aria-describedby="carnet-fare-type-page-heading">
                            <legend className="govuk-fieldset__legend govuk-fieldset__legend--l">
                                <h1 className="govuk-fieldset__heading" id="carnet-fare-type-page-heading">
                                    Select a carnet fare type
                                </h1>
                            </legend>
                            <span className="govuk-hint" id="carnet-fare-type-operator-hint">
                                {operatorName}
                            </span>
                            <FormElementWrapper errors={errors} errorId={errorId} errorClass="govuk-radios--error">
                                <FareTypeRadios fares={carnetRadioProps} />
                            </FormElementWrapper>
                        </fieldset>
                    </div>
                    <input type="submit" value="Continue" id="continue-button" className="govuk-button" />
                </>
            </CsrfForm>
        </TwoThirdsLayout>
    );
};

export const getServerSideProps = (ctx: NextPageContextWithSession): { props: CarnetFareTypeProps } => {
    const carnet = getSessionAttribute(ctx.req, CARNET_FARE_TYPE_ATTRIBUTE);
    if (!carnet || isSchemeOperator(ctx)) {
        if (ctx.res) {
            redirectTo(ctx.res, '/fareType');
        }
    }
    const csrfToken = getCsrfToken(ctx);
    const operatorName = (getSessionAttribute(ctx.req, OPERATOR_ATTRIBUTE) as OperatorAttribute).name || '';
    const fareTypeAttribute = getSessionAttribute(ctx.req, FARE_TYPE_ATTRIBUTE);
    const errors: ErrorInfo[] =
        fareTypeAttribute && isFareTypeAttributeWithErrors(fareTypeAttribute) ? fareTypeAttribute.errors : [];

    return {
        props: {
            csrfToken,
            errors,
            operatorName,
        },
    };
};

export default CarnetFareType;
