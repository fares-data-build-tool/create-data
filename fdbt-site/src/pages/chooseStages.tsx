import React, { ReactElement } from 'react';
import Layout from '../layout/Layout';

const title = 'Choose Stages - Fares Data Build Tool';
const description = 'Choose Stages page of the Fares Data Build Tool';

const ChooseStages = (): ReactElement => (
    <Layout title={title} description={description}>
        <main className="govuk-main-wrapper app-main-class" id="main-content" role="main">
            <form action="/api/chooseStages" method="post">
                <div className="govuk-form-group">
                    <fieldset className="govuk-fieldset" aria-describedby="choose-stages-page-heading">
                        <legend className="govuk-fieldset__legend govuk-fieldset__legend--xl">
                            <h1 className="govuk-fieldset__heading" id="choose-stages-page-heading">
                                How many fare stages does the service have?
                            </h1>
                        </legend>

                        <div className="govuk-form-group">
                            <label className="govuk-hint" htmlFor="width-2">
                                Enter the number of fare stages between 1 - 20 (for example 3)
                            </label>
                            <input
                                className="govuk-input govuk-input--width-2"
                                id="fareStages"
                                name="fareStageInput"
                                type="number"
                                min="1"
                                max="20"
                                maxLength={2}
                                required
                                pattern="^[0-9]*$"
                            />
                        </div>
                    </fieldset>
                </div>
                <input
                    type="submit"
                    value="Continue"
                    id="continue-button"
                    className="govuk-button govuk-button--start"
                />
            </form>
        </main>
    </Layout>
);

export const getServerSideProps = (): {} => {
    return { props: {} };
};

export default ChooseStages;
