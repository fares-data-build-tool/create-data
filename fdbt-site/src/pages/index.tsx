import React, { ReactElement } from 'react';
import TwoThirdsLayout from '../layout/Layout';
import { CustomAppProps, NextPageContextWithSession } from '../interfaces';
import { checkIfMultipleOperators } from '../utils';

const title = 'Create Fares Data';
const description = 'Create Fares Data is a service that allows you to generate data in NeTEx format';

interface StartProps {
    multipleOperators: boolean;
}

const Start = ({ multipleOperators }: StartProps & CustomAppProps): ReactElement => (
    <TwoThirdsLayout title={title} description={description}>
        <h1 className="govuk-heading-xl">Create Fares Data</h1>

        <p className="govuk-body">
            This service is for creating or accessing NeTEx data for public transport services, excluding rail, in
            England. The service can be used by:
        </p>
        <ul className="govuk-list govuk-list--bullet">
            <li>Bus operators running commercial services in England</li>
            <li>Local transport authorities acting on behalf of bus operators</li>
            <li>Local transport authorities that operate their own services</li>
        </ul>

        <p className="govuk-body">Use this service to:</p>
        <ul className="govuk-list govuk-list--bullet">
            <li>Generate fares data in NeTEx format for a new or existing service</li>
            <li>Download your own fares data</li>
        </ul>
        <p className="govuk-body">
            All service data is taken directly from the Traveline National Data Set and the NaPTAN database. To avoid
            data discrepancies, ensure that all service data is correct within these datasets before using the service.
        </p>

        <a
            href={multipleOperators ? '/multipleOperators' : '/fareType'}
            role="button"
            draggable="false"
            className="govuk-button govuk-button--start"
            data-module="govuk-button"
            id="start-now-button"
        >
            Start now
            <svg
                className="govuk-button__start-icon"
                xmlns="http://www.w3.org/2000/svg"
                width="17.5"
                height="19"
                viewBox="0 0 33 40"
                role="presentation"
                focusable="false"
            >
                <path fill="currentColor" d="M0 0h13l20 20-20 20H0l20-20z" />
            </svg>
        </a>
    </TwoThirdsLayout>
);

export const getServerSideProps = (ctx: NextPageContextWithSession): { props: StartProps } => ({
    props: { multipleOperators: checkIfMultipleOperators(ctx) },
});

export default Start;
