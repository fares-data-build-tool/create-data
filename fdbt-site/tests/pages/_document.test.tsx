/* eslint-disable react/jsx-props-no-spreading */
import * as React from 'react';
import { shallow } from 'enzyme';
import { DocumentProps } from 'next/dist/pages/_document';
import MyDocument from '../../src/pages/_document';

describe('_document', () => {
    const props: DocumentProps = {
        html: '',
        __NEXT_DATA__: {
            props: {},
            page: '',
            query: {
                '': '',
            },
            buildId: '',
        },
        dangerousAsPath: '',
        ampPath: '',
        inAmpMode: false,
        hybridAmp: false,
        staticMarkup: false,
        isDevelopment: false,
        devFiles: [''],
        files: [''],
        polyfillFiles: [''],
        dynamicImports: [
            {
                id: '',
                name: '',
                file: '',
                publicPath: '',
            },
        ],
        canonicalBase: '',
        htmlProps: {},
        bodyTags: [{}],
        headTags: [{}],
        lowPriorityFiles: [],
    };

    it('should render correctly', () => {
        const tree = shallow(
            <MyDocument
                {...props}
                nonce=""
                isAuthed
                csrfToken=""
                breadcrumbs={[]}
                url=""
                showCookieBanner
                allowTracking
            />,
        );

        expect(tree).toMatchSnapshot();
    });

    it('should not show the cookie banner when the showCookieBanner attribute is false', () => {
        const tree = shallow(
            <MyDocument
                {...props}
                nonce=""
                isAuthed
                csrfToken=""
                breadcrumbs={[]}
                url=""
                showCookieBanner={false}
                allowTracking
            />,
        );

        expect(tree).toMatchSnapshot();
    });
});
