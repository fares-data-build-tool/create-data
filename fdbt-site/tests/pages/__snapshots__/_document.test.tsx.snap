// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_document should render correctly 1`] = `
<Html
  className="govuk-template app-html-class flexbox no-flexboxtweener"
  lang="en"
>
  <Head
    nonce=""
  />
  <body
    className="govuk-template__body app-body-class js-enabled"
  >
    <Header
      csrfToken=""
      isAuthed={true}
    />
    <div
      className="govuk-width-container app-width-container--wide"
    >
      <AlphaBanner />
      <Main />
    </div>
    <Footer />
    <NextScript
      nonce=""
    />
    <script
      nonce=""
      src="/scripts/all.js"
    />
    <script
      nonce=""
    >
      window.GOVUKFrontend.initAll()
    </script>
  </body>
</Html>
`;
